common/devdoc/sm_requirements.md

```diff
@@ -55,7 +58,7 @@ To achieve the above goals, `sm` maintains the following state:
 Barriers - since they are exclusive - are realized by switching to a state called `SM_OPENED_BARRIER`. Prohibiting regular calls to _begin is achieved by switching temporarily the state from `SM_OPENED` to `SM_OPENED_DRAINING_TO_BARRIER`.
 
 Close is realized by prohibiting all calls (including competing `sm_close_begin` calls) by setting a bit with `InterlockedOr`. `sm_close_begin` will wait for the state to reach `SM_OPENED` and the number of executing calls to be `0`. This allows an ongoing barrier to finish (and return to `SM_OPENED` state), or the executing APIs to finish.
-`sm_close_begin_with_cb` will invoke a callback function between prohibiting all the calls and waiting for the number of executing calls to be `0`. The callback function is useful in various cases, such as, when we need to cancel the ongoing operations, before we wait for the outstanding calls to finish.
+`sm_close_begin_with_cb` will invoke a callback function between prohibiting all the calls and waiting for the number of executing calls to be `0`. The callback function is useful in various cases, such as, when we need to cancel the ongoing operations, before we wait for the outstanding calls to finish.  `sm_close_begin_with_cb` will also invoke an addition callback function (`ON_SM_CLOSING_WILE_OPENING_CALLBACK`) which will let the calling function know if sm is currently in the `opening` state.  This is useful in the scenario where the user will need to do additional steps to close if they are in the opening state.
```

Comment:
>ing state. [](http://example.com/codeflow?start=615&length=10)

I'd maybe just add something like, "It is expected that the user will perform actions to complete the pending open in this callback before returning so that the close can continue."

Just because after the callback returns sm will re-check the state to attempt again to close.